include/deal.II/base/mpi.h

```diff
@@ -708,6 +708,44 @@ namespace Utilities
            const T &          object_to_send,
            const unsigned int root_process = 0);
 
+    /**
+     * Given a partitioned index set space, compute the owning MPI process rank
+     * of each element of a second index set according to the partitioned index
+     * set. A natural usage of this function is to compute for each ghosted
+     * degree of freedom the MPI rank of the process owning that index.
+     *
+     * The user is responsible to partition the input index set space correctly.
+     * Otherwise, there are no limitations: there is no need in partitioning
+     * the index set space continuously. Furthermore, there are no limitations
+     * on the second index set as long as the size matches the first one:
+     * in the case that the second index set also contains locally
+     * owned indices, these indices will be treated correctly and the rank of
+     * this process is returned for those entries.
+     *
+     * Note: This is a collective operation: all processes within the given
+     * communicator have to call this function. Since this function does not
+     * use MPI_Alltoall and MPI_Allgather and instead uses non-blocking
+     * point-to-point communication and only a single non-blocking barrier,
+     * it reduces the memory consumption significantly. This function is suited
+     * for large-scale simulations with >100k MPI ranks.
+     *
+     * @param[in] owned_dofs index set with indices locally owned by this
+     *  process.
+     * @param[in] ghosted_dofs index set containing indices of which the user is
+     *  interested the rank of the owning process.
+     * @param[in] comm MPI communicator.
+     *
+     * @return List of MPI process rank of each entry in the index set
+     *  @p ghosted_dofs. The order coincides with the order within the
+     *  ElementIterator.
+     *
+     * @author Peter Munch, 2019
+     */
+    std::vector<unsigned int>
+    compute_index_owner(const IndexSet &owned_dofs,
+                        const IndexSet &ghosted_dofs,
```

Comment:
I think it might actually be nice to call these arguments `owned_indices` and `indices_to_look_up`. The function is not specific to DoFs, but could be used for any kind of cases that have nothing to do with DoFs.